core/src/main/scala/org/apache/spark/deploy/history/HybridStore.scala

```diff
@@ -0,0 +1,205 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history
+
+import java.io.IOException
+import java.util.Collection
+import java.util.concurrent.atomic.AtomicBoolean
+
+import org.apache.spark.util.kvstore._
+
+/**
+ * A implementation of KVStore that accelerates event logs loading.
+ *
+ * When rebuilding the application state from event logs, HybridStore will
+ * write data to InMemoryStore at first and use a background thread to dump
+ * data to LevelDB once the writing to InMemoryStore is completed.
+ */
+
+private[history] class HybridStore extends KVStore {
+
+  private val inMemoryStore = new InMemoryStore()
+
+  private var levelDB: LevelDB = null
+
+  // Flag to indicate whether we should use inMemoryStore or levelDB
+  private[history] val shouldUseInMemoryStore = new AtomicBoolean(true)
+
+  // A background thread that dumps data from inMemoryStore to levelDB
+  private var backgroundThread: Thread = null
+
+  // Objects of these classes will be dumped to levelDB in the background thread
+  private var klassList: Seq[Class[_]] = List(
```

Comment:
This looks to be fragile. Is there a reason we have to apply allow-list instead of dumping all of the entities?

Reply:
I think the main purpose is to handle CachedQuantile specifically. CachedQuantile is not on this list, so it won't be dumped to levelDB in the background thread. Right now my strategy is to write CachedQuantile to inMemoryStore and levelDB simultaneously during the transition process (background thread is alive), this strategy seems ok in my manual test. 

CachedQuantile is quite special, it will only be added to kvstore after rebuildAppStore() finishes. that means if we don't handle it specially it will be added to inMemoryStore when the background thread is dumping data(including CachedQuantile objects) to leveldb and hence cause conflicts.

Reply:
This requires change for every time when any new api class is introduced, which is easy to miss - that's why I said 'fragile'. If we have to have some kind of allow-list vs reject-list, I think reject-list is better, because storing something after replaying is an exceptional case.

Actually I don't think they even need to be retained at all during transition - it's optional to migrate them to the level DB. All of them should be just some sort of cache. The status of KV store is stable after finished replaying, otherwise it would be problematic for in-progress application. Think about in-progress application - KV store becomes invalidated once the event log file is updated which means newly stored entries are lost. In practice it doesn't matter at all.

Reply:
ok, I will use a reject-list and add CachedQuantile to reject-list. And I think when using reject-list, we need a HashSet to record the classes that has been added to kvstore.

yeah, I agree it's optional to retain these CachedQuantile objects.

Reply:
If we don't write CachedQuantile to levelDB during the transition process, we might be able to get rid of reject list or allow list, and hence no need to handle some classes specially. I will test that idea and push the code changes later.

Reply:
If possible and works properly (though I guess so), it would be ideal if we don't specify the case "CachedQuantile" but don't apply all write operations to Level DB during the transition. It would trigger one more calculation/storing of CachedQuantiles after the transition to Level DB, but IMHO it wouldn't matter much as we already saved lots of seconds on loading and the overhead should be relatively small enough.

Reply:
agree

Reply:
Explicit logic for handling particular classes are removed.